### DISCUSSION TITLE
Can we have xpaths for segments?

### DISCUSSION THREAD

**Discussion by fernandocamargoai** [FitLayout/FitLayout#18] [General]
Hello. I've just discovered this project when looking for ways to segment pages. I tried it a little using the docker image with CLI and liked the results with VIPS so far. However, one thing I missed is how to relate the segments to the original HTML. I'd like to have XPaths, CSS selectors or any other way to relate the segments to the original HTML. Would that be possible?

**Comment by radkovo** [comment]
Hello. This is possible but it's not completely straightforward for the following reasons:

1. FitLayout operates on the rendered page that is represented as a tree of *boxes*, which do not exactly correspond to HTML elements. In fact, during rendering by the rendering engine, each HTML element can generate 0 to N boxes displayed on the output page (e.g. a paragraph that contains multiple text lines). All the boxes generated by a single HTML element have the same source XPath but they can (in theory) appear in different visual segments (but most probably they won't, at least in VIPS).
2. The result of segmentation using VIPS is a tree of *visual areas* (segments) where the leaf areas contain 1 to N boxes (by applying visual separators etc.) So the detected segment does not necessarily correspond to a single HTML element; often, it contains multiple boxes that can be mapped to their source HTML elements as described above.

So, to get the links to HTML elements, you need to

1. Get the list of boxes that belong to every (leaf) segment in the area tree.
2. For every such box, obtain the XPath that identifies the source HTML element.

When using the CLI, the XPaths are not part of the data export in the XML and JSON formats because the [default export implementation](https://github.com/FitLayout/FitLayout/blob/main/fitlayout-io/src/main/java/cz/vutbr/fit/layout/io/ArtifactStreamOutput.java) only exports the most common attributes (but I can include the XPaths easily if you find it useful). At the moment, the best option is to use the export to the Turtle (ttl) format that contains all the attributes. The CLI pipeline could look for example like this:

```bash
./fitlayout.sh \
    RENDER -b puppeteer https://cssbox.sf.net \
    EXPORT -f turtle -o boxes.ttl  \
    SEGMENT -m vips -O pDoC=9 \
    EXPORT -f turtle -o areas.ttl
```

This renders the page and stores the information about the source boxes to `boxes.ttl`. Then, the VIPS segmentation is performed and the result is exported to `areas.ttl`. The exported areas contain links to the individual boxes that are contained in the areas.

Alternatively, you can avoid using the CLI and use directly the Java API that gives access of all the attributes of the areas. For example, you can start with the [VIPS segmentation demo](https://github.com/FitLayout/Demos/blob/891148ffdade06e55ac22405b509f80ec0aa8516/src/main/java/cz/vutbr/fit/layout/demo/segmentation/SegmVIPS.java#L59) and replace the default export using the `XMLOutputOperator` by your custom export.

**Comment by fernandocamargoai** [comment]
Thank you very much, @radkovo!
